### PR TITLE
model: Add support for Motif 3 Beta 

### PR DESCRIPTION
**Pull request by timkhronos**
## Overview
Adds support for Motif 3 Beta. 

### GDLA (Grouped Differential Latent Attention)
MLA-style low rank projections, but with 16 GQA KV heads. ~~It is implemented as plain GQA with build_attn(), and not through an MLA cache.~~ On top of the attention output, heads form 16 groups of (4 signal + 1 noise), plus an optional sigmoid output gate.

The full attention layers now use an MLA latent cache, lowering context memory usage by 8.9x, allowing the model's full trained context to fit in 4.5GiBs at F16. The SWA layers keep their standard GQA representation.

### Grouped PolyNorm activation

Used instead of SiLU, with per expert coefficients for routed experts, and distinct clamp behaviour for dense MLPs vs experts. MoE routing itself is similar to DeepSeek V3(sigmoid top k), with the difference being the activation between the expert gate/up and down matmuls.

### mHC (Manifold-constrained Hyper-Connections) 

Structurally similar to the existing Deepseek V4 hyper connections, so the graph reuses the fused DSV4 HC ops for steam reduction, combination and Sinkhorn, with the unfused as fallback. The motif specific pieces (sigmoid pre/post gates,post gate coefficient, and flattened four stream rms normalization) are computed in the graph. The fused path reduces graph nodes from 40k to 10k, speeds up prefill by about 70% and decode by around 30%.

### Interleaved SWA + Yarn

Full attention on every 4th layer(layer_index % 4 == 0). Swa layers use plain Rope. Reuses the existing iswa Kv cache. Motif's Yarn never scales cos/sin. The mscale^2 factor is folded into the softmax only for the full attention layers, so the graph cancels ggml's internal cos/sin factor. 

### Conversion

Explicit tensor map. MTP heads are skipped during conversion. Sensitive tensors (router, expert bias, PolyNorm, mHC and the GDLA attention projections) are forced to F32. Vocab is loaded without transformers, since the tokenizer/model repository was serialized by transformers v5(tokenizer_class: TokenizersBackend, list-valued extra_special_tokens), which the pinned version cannot instantiate.

GGUFs converted before the MLA change will still load for now, however will be unable to use the memory efficient MLA cache. Either reconvert fresh (run hf to gguf again) or get .v3 ggufs from the linked repo below when they finish uploading.

### Verification

A tiny fp32 dummy model with all features vs HF. prefill logits match at 1e-6 to 1e-4 over all positions, and decode matches single batch prefill.
An independent numpy transcription of the transformers math agrees with the llama graph at ~5e-8, and I verified the SWA mask offset against explicit masking with a negative control (off by one mask failed the test by >300x)

The real 314B model converts, loads and generates coherently up to 32k (highest tested due to lack of flash attention) at Q4_K_M. 

Converted GGUFs are available [here](https://huggingface.co/avar6/Motif-3-Beta-gguf)(Q2-Q8) and [here](https://huggingface.co/Serpen/Motif-GGUF)(Q5 and BF16) 

### Limitations

Cuda flash attention depends on #26404 and is only available with devices that have turing_mma_availabe, since the generic tile kernel lacks some of the safety features the MMA kernel has, therefore it miscomputed significantly.  

Without FA, the KQ buffer at large n_ubatch * n_ctx sizes is big, so lower -ub sizes are recommended for now until that PR get's merged.

As previously mentioned MTP heads are not covered here.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES, Opus and Fable were used to figure out several transformers 5 conversion issues, subsequently used to help refactor motif.py with the findings. Used to help plan out the best way to implement the model's custom components, to write the validation tests. More specifically, the converter had medium amounts of AI assistance, including the mentioned transformers v5 issues, while the implementation had small amounts of assistance especially in regards to the GDLA attention code. Additionally helped validate the MLA absorption math. Every change was reviewed and validated by me.

### CODE REVIEW

**Comment by EugeoSynthesisThirtyTwo** [comment]
Hi, thank you for trying to implement Motif ! I tried your fork, and Q3_K_M quickly breaks and repeats itself endlessly.
Q4_K_M is way better and manages to stay coherent, but it looks like it has attention troubles? It looks like it keeps correcting itself, I don't know if it's the base model who is like that, it's hard to tell since I can't test it anywhere else

I start with the following commands

Q3_K_M is entirely on the CPU
`llama-server.exe -m Motif3-Q3_K_M.gguf-00001-of-00004.gguf --n-gpu-layers 0 --port 5001 -c 32768 --host 0.0.0.0 -fa off --cpu-range 16-31 --threads 16 -ub 128 -b 128 --checkpoint-min-step 2048 --swa-checkpoints 32 --no-mmap -np 1`

Q4_K_M couldn't fit in my RAM so I had to offload some layers, I don't know if it changes the result
`llama-server.exe -m Motif3-Q4_K_M.gguf-00001-of-00004.gguf --n-gpu-layers 10 --port 5001 -c 32768 --host 0.0.0.0 -fa off --cpu-range 16-31 --threads 16 -ub 128 -b 128 --checkpoint-min-step 2048 --swa-checkpoints 32 --no-mmap -np 1`

**Comment by timkhronos** [comment]
@EugeoSynthesisThirtyTwo Hi! If you have only tried the original GGUFs, they have had an issue with some critical tensors(including attention tensors) being quantized aggresively. If you re-download them then they should hold up much better at the lower quants I believe. (they were updated around 12 hours ago I believe)

Do note that they'll need to be requanted again, as there are still a few tensors left a bit too low, but they should already work much better if you originally tried the first batch. [here](https://huggingface.co/avar6/Motif-3-Beta-gguf). 

I also have a Q5 uploaded [here](https://huggingface.co/Serpen/Motif-GGUF/tree/main) that I would consider an 'ideal' quant, but based on what you said you might not be able to fit it. My repo also has the converted BF16 GGUF files if you want to make your own quant.

I think that was the big contributor to your issues. Gpu offloading is perfectly fine, it's only Flash Attention that doesn't work. If you are using fresher ggufs I'll look into what might be the issue, as I'd expect those to hold up even at Q3.

**Comment by EugeoSynthesisThirtyTwo** [comment]
Hi, I checked the SHA256 and I can confirm that my files are from your second batch

By the way, I have 192 GB of ram, and 2x32 GB of vram, I haven't tried the Q5 yet, but maybe it can fit

**Comment by timkhronos** [comment]
@EugeoSynthesisThirtyTwo Yeah, my Q5 is around ~230Gb so you should be able to fit it comfortably.  It's also possible the model just quantizes poorly? For what it's worth the smaller quants in Avar's repos have been updated as well.
